sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -702,9 +702,13 @@ abstract class TypeCoercionBase {
         }.getOrElse(b)  // If there is no applicable conversion, leave expression unchanged.
 
       case e: ImplicitCastInputTypes if e.inputTypes.nonEmpty =>
-        val children: Seq[Expression] = e.children.zip(e.inputTypes).map { case (in, expected) =>
+        val children: Seq[Expression] = e.children.zip(e.inputTypes).map {
+          case (expr: Expression, StringType) if expr.dataType.isInstanceOf[StringType] => expr
+          case (expr: Expression, StringTypeBinary | StringTypeBinaryLcase | StringTypeAnyCollation)
+            if !expr.dataType.isInstanceOf[StringType] =>
+              implicitCast(expr, StringType).getOrElse(expr)
```

Comment:
We should put the fix in `def implicitCast`. Otherwise it looks hacky to make a special case here.